examples/01_mvp/01_mvp.cpp

```diff
@@ -0,0 +1,193 @@
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+// This example demonstrates how to perform a distributed matrix-vector product (A * x = y)
+// using KokkosComm. The matrix A is distributed among the ranks by blocks of contiguous rows.
+// Each rank owns a part of the vector x and will communicate it with other ranks step by step.
+// At each step a node communicates with two other nodes to send and receive data.
+
+// Two helper objects are used to manage the distributed work:
+// - RankDims: contains the number of rows, the start and end row indices for each rank
+// - RankCommInfo: contains the ranks to send data to and receive data from
+
+// Other algorithms can be used to distribute the work.
+
+#include <KokkosComm/KokkosComm.hpp>
+#include <iostream>
+
+struct RankDims {
+  unsigned int nb_rows;
+  unsigned int row_start;
+  unsigned int row_end;
+
+  RankDims(long N, int rank, int size) {
+    this->nb_rows   = N / size;
+    this->row_start = rank * this->nb_rows;
+    this->row_end   = (rank + 1) * this->nb_rows;
+
+    if (rank == size - 1) {
+      this->row_end = N;
+      this->nb_rows = this->row_end - this->row_start;
+    }
+  }
+};
+
+struct RankCommInfo {
+  Kokkos::View<unsigned int*> recv_ranks;
+  Kokkos::View<unsigned int*> send_ranks;
+
+  RankCommInfo(int rank, int size) {
+    this->recv_ranks = Kokkos::View<unsigned int*>("recv_ranks", size);
+    this->send_ranks = Kokkos::View<unsigned int*>("send_ranks", size);
+
+    for (int i = 0; i < size; i++) {
+      this->recv_ranks(i) = (rank + i) % size;
+      this->send_ranks(i) = (rank - i + size) % size;
+    }
+  }
+};
+
+// Provide a default value for N
+void checkArgs(long& N) {
+  if (N < 1) {
+    N = 1 << 12;
+  }
+}
+
+template <typename View>
+bool verify_result(View& y, long N) {
+  for (unsigned long i = 0; i < y.extent(0); i++) {
+    if (y(i) != N) {
+      return false;
+    }
+  }
+  return true;
+}
+
+int main(int argc, char* argv[]) {
+  long N = -1;
+
+  for (int i = 0; i < argc; i++) {
+    if (strcmp(argv[i], "-N") == 0 && i + 1 < argc) {
+      N = std::atoi(argv[++i]);
+    }
+    if (strcmp(argv[i], "-h") == 0) {
+      std::cout << "KokkosComm dense square matrix-vector product example \n"
+                << "  Usage: " << argv[0] << " [-N <size>] default size is 2^12" << std::endl;
+      return 0;
+    }
+  }
+  checkArgs(N);
+
+  int provided;
+  MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+  if (provided != MPI_THREAD_MULTIPLE) {
+    throw std::runtime_error("MPI_THREAD_MULTIPLE is needed");
+  }
+
+  Kokkos::initialize(argc, argv);
+  {
+  using ExecSpace   = Kokkos::DefaultExecutionSpace;
+  using CommSpace   = KokkosComm::DefaultCommunicationSpace;
```

Comment:
Since this example relies on MPI-specific functions (`KokkosComm::mpi::barrier`), I would use the MPI comm space explicitly:
```suggestion
  using CommSpace   = KokkosComm::Mpi;
```

Reply:
done